[processor/attributesprocessor] Add metric support

crobert-1 opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
This change adds metric support to the attributes processor. Metric data point attributes can now be modified in all of the same ways as spans and logs.

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

Included unit tests, also did manual testing. Manual tests show that insert, delete and extract were all working as expected.
**Documentation:** <Describe the documentation added.>

README was updated, as well as extra comments in shared attributes processor code.

crobert-1 committed:
Add metric support to the attributes processor

This change adds metric support to the attributes processor. Metric
data point attributes can now be modified in all of the same ways
as spans and logs. This change also adds testing that is all passing.

crobert-1 committed:
Update changelog with PR reference

crobert-1 committed:
Fix linting failure of improper import grouping

crobert-1 committed:
Fix linting failures of incorrect indentation

pmcollins commented on internal/coreinternal/processor/filtermetric/filtermetric.go:
```diff
@@ -16,17 +16,57 @@ package filtermetric // import "github.com/open-telemetry/opentelemetry-collecto
 	MatchMetric(metric pdata.Metric) (bool, error)
 }
 
+func CreateMatchPropertiesFromDefault(properties *filterconfig.MatchProperties) *MatchProperties {
```
Should this live next to the MatchProperties struct?

pmcollins reviewed (commented):
Fine work -- just a few questions.

tigrannajaryan commented on internal/coreinternal/processor/filterconfig/config.go:
```diff
@@ -24,32 +24,32 @@ import (
+	// Include specifies the set of span/log/metric properties that must be present in order
 	// for this processor to apply to it.
-	// Note: If `exclude` is specified, the span/log is compared against those
+	// Note: If `exclude` is specified, the span/log/metric is compared against those
```
To avoid repeating `span/log/metric` everywhere it may be better to instead say `input data`, e.g.:
```suggestion
	// Note: If `exclude` is specified, input data is compared against those
```
Similarly in other places.

pmm-sumo replied:
This suggestion is perhaps a nit but it seems that it was dismissed (i.e. no change applied here or elsewhere) and the thread was resolved without a comment

crobert-1 replied:
My apologies, you're correct, I'm going through and resolving when I change them locally vs. when I commit. I should probably add comments confirming that I'll do what's requested though.  Thanks for pointing that out!

Just to confirm I'll be changing the span/log/metric phrasing to match what Tigran's requested as "input data".

tigrannajaryan replied:
It is best to resolve after the change is visible here in Github otherwise the reviewers have no way of confirming that it is done. It is also useful to comment on the issue saying what you did to resolve.

crobert-1 commented on processor/attributesprocessor/README.md:
```diff
@@ -169,6 +176,18 @@ attributes:
+
+      # Expressions are used to match metrics. A match occurs if the expression matches at least
+      # one data point within a metric.
+      expressions: [<item1>, ..., <itemN>]
```
I've added a link, but couldn't figure out how to format it better in the YAML block of a readme, so it'll just show up as a plain link.

pmm-sumo commented on internal/coreinternal/processor/filterconfig/config.go:
```diff
@@ -107,6 +115,14 @@ type MatchProperties struct {
 	// This is an optional field.
 	Libraries []InstrumentationLibrary `mapstructure:"libraries"`
+
+	// Expressions specifies the list of expr expressions to match metrics against.
+	// A match occurs if any datapoint in a metric matches at least one expression in this list.
+	Expressions []string `mapstructure:"expressions"`
```
+1 to that, also if this functionality is really needed probably it should be extracted into separate PR (so one adds metrics, the other for expressions)

crobert-1 replied:
I added expressions as a possible matching functionality as the metric filter already has the expression matching implemented. I assumed since it was a simple change it made sense to enable it.

From your feedback it looks like it adds more confusion to the customer than any benefit it would add, so I'll remove it for now (since there isn't a specific requirement to support expression matching at this point, AFAIK).

pmm-sumo commented on internal/coreinternal/processor/filterconfig/config.go:
```diff
@@ -107,6 +115,14 @@ type MatchProperties struct {
+	// A match occurs if any datapoint in a metric matches at least one expression in this list.
+	Expressions []string `mapstructure:"expressions"`
+
+	// ResourceAttributes defines a list of possible resource attributes to match metrics against.
+	// A match occurs if any resource attribute matches all expressions in this given list.
+	ResourceAttributes []Attribute `mapstructure:"resource_attributes"`
```
+1 again, I am confused how does it differs from [resourceprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/resourceprocessor)

If we want more complex logic then I believe we're going to move towards [transformprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/transformprocessor)

crobert-1 replied:
I'll remove the new attribute ResourceAttributes and use the existing Resource attribute.

pmm-sumo commented on processor/attributesprocessor/README.md:
```diff
@@ -169,6 +176,18 @@ attributes:
+
+      # Resource attributes defines a list of possible resource attributes to match metrics against.
+      # A match occurs if any resource attribute matches all expressions in this given list.
+      resource_attributes: [<item1>, ..., <itemN>]
```
Additional problem with this parameter is that it works only for metrics but is named like a universal one

crobert-1 replied:
I've removed this in favor of using the pre-existing resource attribute.

crobert-1 commented on internal/coreinternal/processor/filterconfig/config.go:
```diff
@@ -136,6 +152,9 @@ func (mp *MatchProperties) ValidateForLogs() error {
 	return nil
 }
 
+// ValidateForMetrics validates properties for metrics
+// Need to implement method, add comments to new members of MatchProperties
+
```
I left this as a TODO and missed it while committing. I'm adding the missing method in the next commit.

crobert-1 commented on internal/coreinternal/processor/filterconfig/config.go:
```diff
@@ -107,6 +115,14 @@ type MatchProperties struct {
 	// A match occurs if the span's implementation library matches at least one item in this list.
 	// This is an optional field.
 	Libraries []InstrumentationLibrary `mapstructure:"libraries"`
```
Would that be considered an official requirement for metric support here? Library matching isn't currently implemented by the metric filter, so I'm not sure if it makes more sense to add that as a separate feature/enhancement request, or add that functionality with this review.

dmitryax replied:
Ok we can keep it, out of the scope of this PR

crobert-1 committed:
Changes requested in review

- Remove out-dated comments
- Change all span/log/metric references to "input data"
- Add comments about usage of attributes processor vs.
metric transform processor.

crobert-1 commented on processor/attributesprocessor/README.md:
```diff
@@ -1,21 +1,21 @@
 # Attributes Processor
```
I've added a description comparing functionalities, along with recommendations on when to use them.

crobert-1 commented on internal/coreinternal/processor/filtermetric/filtermetric.go:
```diff
@@ -16,17 +16,57 @@ package filtermetric // import "github.com/open-telemetry/opentelemetry-collecto
+	if include != nil {
+		// A false (or an error) returned in this case means the metric should not be processed.
+		i, err := include.MatchMetric(metric)
+		if !i || err != nil {
```
Is there a specific log library or format required here? Most other places I've seen in OpenTelemetry are passing around a `TelemetrySettings` object, or something similar.

Is it acceptable here to just add some generic logger with a warning message?

pmcollins replied:
Where `NewMatcher` is called, there's a logger available in the `component.ProcessorCreateSettings` arg at the path `processorCreateSettings.TelemetrySettings.Logger` that could be passed into `NewMatcher`, but if we're not using expr matching then there may not be a need. Will resolve this thread.

crobert-1 committed:
Merge branch 'main' into metric_attribute_processor

dmitryax commented on internal/coreinternal/processor/filtermetric/filtermetric.go:
```diff
@@ -25,8 +25,36 @@ type Matcher interface {
+	}
+
+	if exclude != nil {
+		// A true (or an error) returned in this case means the span should not be processed.
```
I think we need consistent behavior in case of error. Looks like we will not skip metrics in case of error in include matcher, but skip it in case of exclude matcher error.

Probably we should skip the metric if an error happens with any matcher. Ideally we should log the error with debug level. To do that, `Logger` from the processor instance should be passed here .

crobert-1 replied:
From my understanding the metric is always skipped if an error is hit. When an error is hit in include we skip, and when it's hit in exclude we skip.

I agree about logging though, so I'll add a log message when we're skipping due to error.

dmitryax commented on processor/attributesprocessor/README.md:
```diff
@@ -116,16 +116,49 @@ processors:
 Refer to [config.yaml](./testdata/config.yaml) for detailed
 examples on using the processor.
 
+### Attributes Processor for Metrics vs. Metric Transform Processor
```
Nice! Thanks for describing the difference

dmitryax commented on processor/metricstransformprocessor/README.md:
```diff
@@ -311,3 +311,29 @@ operations:
+
+### Attributes Processor for Metrics vs. Metric Transform Processor
+
+Regarding metric support, these two processors have overlapping functionality. They can both do simple modifications
```
I think a lint to the Attributes Processor would be useful here

crobert-1 replied:
I'll add a link

dmitryax commented on processor/attributesprocessor/README.md:
```diff
@@ -116,16 +116,49 @@ processors:
 
+### Attributes Processor for Metrics vs. Metric Transform Processor
+
+Regarding metric support, these two processors have overlapping functionality. They can both do simple modifications
```
Same here, link to the Metric Transform Processor would be nice to add

crobert-1 replied:
I'll add a link here as well, thanks for the suggestion.

dmitryax commented on processor/metricstransformprocessor/README.md:
````diff
@@ -311,3 +311,29 @@ operations:
   group_resource_labels: {"resouce.type": "container", "source": "kubelet"}
 ```
+
+### Attributes Processor for Metrics vs. Metric Transform Processor
````
```suggestion
### Metric Transform Processor vs Attributes Processor for Metrics
```
?

crobert-1 replied:
That is a much better way to word it, I'll update.

dmitryax committed:
Merge branch 'main' into metric_attribute_processor

crobert-1 committed:
Changes requested in PR

- Add logging when skipping metrics
- Add links to different processors

crobert-1 committed:
Linting fixes

crobert-1 committed:
Merge remote-tracking branch 'origin/main' into metric_attribute_processor

dmitryax reviewed (approved):
LGTM

codeboten reviewed (commented):
@pmm-sumo @pmcollins @tigrannajaryan please take a look and approve if your comments have been addressed
